### ISSUE TITLE
Update the code to 2021 practices?

### ISSUE THREAD

**Issue by greggman** [esangel/WebGL#11]
Would you consider a series of PRs that update the code to slightly more modern practices? Some are minor, some less so.

Examples might include

* Remove all uses of `var` for `let` and `const`

   `var` is basically considered bad practice at this point

* no need for `type="text/javascript"

    this is not needed for script tags

* stop using `window.onload`

    Just move your script tags to the bottom or use `defer`

* consider using es6 modules for loading scripts

* use `requestAnimationFrame` instead of `requestAnimFrame`
   
   There's no need for the polyfill in 2021

* put the shaders in multiline template strings

   They've been in every browser since 2010 or earlier

* get rid of "Oops, your browser doesn't support canvas"

   There are no browser that don't support canvas in 2021

**Comment by esangel** [comment]
Gregg,

Thanks for your input. I’m agreement in principle with almost all your examples.  Let me address the issues in dealing with them in the context of my and Dave Shreiner’s textbooks.

The largest part of our audience and the ones we write for are students in computer science and engineering programs. Almost none of these programs use the JavaScript API in any of their core courses. Although I personally am troubled by this reality, the problem for us is when students get to a course on computer graphics, the majority of them have never programmed with JS. Because the instructors’ goal is teaching computer graphics using OpenGL/WebGL and not creating experts in JS, we decided to stick with a very simple subset of JS, even if it meant our code is often not very efficient or JS-like or up to the present versions of JS.

Note that most of your suggestions require ES6. When we discussed our plans for the 8th Edition of Interactive Computer Graphics (Pearson Education) which came out last year, we decided to move to WebGL 2.0 but, given the background of our readers, we decided to stick with ES5. If there is a 9th Edition, we’ll almost definitely move to at least ES6. For now, we can do two things. We can clean up the existing code on our website, which will still be ES5. We can also create ES6 versions of the present examples and put them on our website. The latter will take a little while.

Ed


_______________________

Ed Angel

Founding Director, Art, Research, Technology and Science Laboratory (ARTS Lab)
Professor Emeritus of Computer Science, University of New Mexico

1017 Sierra Pinon
Santa Fe, NM 87501
505-984-0136 (home)		 	angel@cs.unm.edu <mailto:angel@cs.unm.edu>
505-453-4944 (cell) 				http://www.cs.unm.edu/~angel <http://www.cs.unm.edu/~angel>

**Comment by greggman** [comment]
> Note that most of your suggestions require ES6.

FYI, only 2 of my suggestions (import, no var) needs es6. All the rest are normal stuff from years ago

**Comment by esangel** [comment]
True. I should have been more careful to note the version of ES5. 

The main issue in updating the examples to use let and const  is the change in scoping rules. 

Ed
_______________________

Ed Angel

Founding Director, Art, Research, Technology and Science Laboratory (ARTS Lab)
Professor Emeritus of Computer Science, University of New Mexico

1017 Sierra Pinon
Santa Fe, NM 87501
505-984-0136 (home)		 	angel@cs.unm.edu <mailto:angel@cs.unm.edu>
505-453-4944 (cell) 				http://www.cs.unm.edu/~angel <http://www.cs.unm.edu/~angel>

**Comment by greggman** [comment]
`let` and `const` would be a big change, I agree

I hope this isn't too pushy, But, for example, searching and replacing `requestAnimFrame` with `requestAnimationFrame` is a small change and would stop teaching students about some non-existent function outside the class.

That seems like a minor change?

Similar, searching and replacing ` type="text/javascript"` with nothing is a minor change that stops repeating outdated practices.

The `Oops` just removes clutter and is also outdated

Maybe there is other work todo but at a glance those seem like 5 minute changes that have some positive benefits though maybe minor.

Anyway, I'll shut up now. I'm just the guy that answers your student's questions on stack overflow every few months when a new batch comes through :P

**Comment by esangel** [comment]
Where are the students looking at the examples? Are they using interactivecomputergraphics,com and the code for the most recent edition (8E)? I removed all the references to requestAnimFrame quite a while ago. We’ve had a lot of trouble with illegal copies of older editions on the internet. One of the consequences is that students who use them are using legacy versions of the code.  I’ll check to see if I missed one or two. As for the others, I agree and will do. 

Ed
_______________________

Ed Angel

Founding Director, Art, Research, Technology and Science Laboratory (ARTS Lab)
Professor Emeritus of Computer Science, University of New Mexico

1017 Sierra Pinon
Santa Fe, NM 87501
505-984-0136 (home)		 	angel@cs.unm.edu <mailto:angel@cs.unm.edu>
505-453-4944 (cell) 				http://www.cs.unm.edu/~angel <http://www.cs.unm.edu/~angel>

**Comment by greggman** [comment]
I have no idea where the students are getting the materials. They just show up on stackoverflow.
You're right they might be getting old materials from somewhere else.
